docs/docs/reference/custom-components/02-custom-component.md

```diff
@@ -0,0 +1,647 @@
+class CppComponent : public modulo_components::Component {
+public:
+  explicit CppComponent(const rclcpp::NodeOptions& options);
+
+private:
+  std::shared_ptr<double> output_number_;
+  std::shared_ptr<state_representation::CartesianPose> output_pose_;
+};
+```
+
+```cpp title="src/CppComponent.cpp"
+CppComponent::CppComponent(const rclcpp::NodeOptions& options):
+    modulo_components::Component(options),
+    output_number_(std::make_shared<double>(3.14),
+    output_pose_(std::make_shared<state_representation::CartesianPose>()) {
+  // ...
+  
+  // the message type is inferred from the data type
+  this->add_output("number", this->output_number_);
+  this->add_output("pose", this->output_pose_);
+}
+```
+
+</TabItem>
+</Tabs>
+
+The value of the class attribute can then be freely changed in other parts of the component implementation to update the
+published value. For example, it may be changed in a periodic step function, or as a result of parameter validation, a
+service call, an input subscription callback or a lifecycle transition.
+
+:::note
+
+Empty states will not be published.
+
+Lifecycle components publish outputs only when the component is in the `ACTIVE` lifecycle state.
+
+:::
+
+<!-- TODO
+## Adding services
+-->
+
+<!-- TODO
+## Adding predicates
+-->
+
+## Adding periodic behavior
+
+Every component has a built-in `period` parameter that defines the execution rate of periodic behaviors in seconds. For
+example, a component with a period of 0.1 seconds will execute periodic behaviors ten times per second. The evaluation
+of periodic behaviors is referred to as the component "step".
+
+:::note
+
+Lifecycle components evaluate periodic callback functions only when the component is in the `ACTIVE` lifecycle state.
+
+:::
+
+Custom periodic behavior can be implemented by overriding the `on_step_callback` function. For example, a random number
+generator component might run a pseudo-random function every step so that the published output is always updated.
+
+:::tip
+
+Custom periodic callback functions are always evaluated _before_ outputs are published. If a periodic callback function
+makes any changes to the value of a data object that is bound to and output, the latest value will be published in the
+same component step.
+
+:::
+
+
+<Tabs groupId="programming-language">
+<TabItem value="python" label="Python">
+
+```python title="custom_component.py"
+from random import random
+
+from std_msgs.msg import Float64
+```
+
+```python title="custom_component.py"
+    def __init__(self, node_name: str, *args, **kwargs):
+        # ...
+        
+        self._random_number = random()
+        self.add_output("random_number", "_random_number", Float64)
+    
+    def on_step_callback():
+        self._random_number = random()
+```
+
+</TabItem>
+<TabItem value="c++" label="C++">
+
+```cpp title="include/custom_component_package/CppComponent.hpp"
+class CppComponent : public modulo_components::Component {
```

Comment:
This is not correct, is it? Components don't have an on_step_callback. https://github.com/aica-technology/modulo/blob/main/source/modulo_components/src/Component.cpp

Reply:
I wrote this preemptively because I think they should, I was planning to tag you in this when you got back.
https://github.com/aica-technology/modulo/issues/49